sandbox/apparmor/apparmor.go

```diff
@@ -56,6 +57,221 @@ func ValidateNoAppArmorRegexp(s string) error {
+// /{usr[^/],var[^/]}** r,
+// /{usr/[^l],var/[^l]}** r,
+// /{usr/l[^i],var/l[^i]}** r,
+// /{usr/li[^b],var/li[^b]}** r,
+// /{usr/lib[^/],var/lib[^/]}** r,
+// /{usr/lib/[^s],var/lib/[^s]}** r,
+// /{usr/lib/s[^n],var/lib/s[^n]}** r,
+// /{usr/lib/sn[^a],var/lib/sn[^a]}** r,
+// /{usr/lib/sna[^p],var/lib/sna[^p]}** r,
+// /{usr/lib/snap[^d],var/lib/snap[^d]}** r,
+//
+// Note that with the previous rules, /usr/lib/snapdaemon would not match any rule
+//
+// This function has the following limitations:
+// The first character after a subpattern common to two or more excludePatterns cannot
+// be '*' on any of the excludePatterns that share the common prefix.
+// Eg. ["/snap/core/**", "/snap/*/core/**"] where '/snap/' would be the subpattern common
+// to both excludePattern, and '*' would be the first character after the common subpattern
+// in the second excludePattern.
+// This is because there are no apparmor rules that can fulfill both requirements.
+// For /snap/[^c]** -> It will also match /snap/a/core/** which should be excluded by
+// the second pattern.
+// For /snap/[^c*]** -> It will also exclude access to /snap/a/a that should be allowed
+// as it is not explicitly excluded by any pattern
+//
+// When the '*' is used to exclude suffixes, like in ["/*.bin"], rules should be generated
+// in a reverse way:
+//
+// /*[^n]{,/**} rw,
+// /*[^i]n{,/**} rw,
+// /*[^b]in{,/**} rw,
+// /*[^.]bin{,/**} rw,
+//
+// While generating those rules is technically possible, it will make the logic way more
+// complex, thus the function would just return an error if a pattern of this kind is found.
+// This functionality can be added in a subsequent PR if needed in the future
+func GenerateAAREExclusionPatterns(excludePatterns []string, opts *AAREExclusionPatternsOptions) (string, error) {
+	if len(excludePatterns) == 0 {
+		return "", fmt.Errorf("no patterns provided")
+	}
+	seen := map[string]bool{}
+	for _, patt := range excludePatterns {
+		// check for duplicates
+		if seen[patt] {
+			return "", errors.New("exclude patterns contain duplicates")
+		}
+		seen[patt] = true
+
+		// check if it is at least legnth 2
+		if len(patt) < 2 {
+			return "", errors.New("exclude patterns must be at least length 2")
+		}
+
+		// check that it starts as an absolute path
+		if patt[0] != '/' {
+			return "", errors.New("exclude patterns must be absolute filepaths")
+		}
+
+		// TODO: should we also validate that the only character in the pattern
+		// from AARE is "*" ?
+	}
+	if opts == nil {
+		opts = &AAREExclusionPatternsOptions{}
+	}
+	return generateAAREExclusionPatternsGenericImpl(excludePatterns, opts)
+
+}
+
+func generateAAREExclusionPatternsGenericImpl(excludePatterns []string, opts *AAREExclusionPatternsOptions) (string, error) {
+
+	// Find the length of longest pattern (size)
+	size := 0
+	for _, pattern := range excludePatterns {
+		if len(pattern) > size {
+			size = len(pattern)
+		}
+	}
+
+	// Find the longest prefix common to ALL patterns.
```

Comment:
Please break this out to a helper function. It will make the rest more readable and shorter. I think it can go to the `strutil` package. Write a separate test for this please.

Reply:
Done in commit commit c7b920189797ba12ab53f47a5e3b8da1e01f162e